Added design document for remote and multi-machine launching (#297)
ros2/design: articles/153_roslaunch_remote_launch.md

# Multi-Machine and Remote Launching
## Goals
- Connect to a remote host and run nodes on it

>     - Ensure this capability does not compromise security

Reviewer: We're currently working on launch integration for security. In order to make sure it scales to remote launching, we might want to consider ways to send security files over to the remote machine.

- Support arbitrary remote execution methods
- Monitor the status of and manage lifecycles of nodes across systems
- Shutdown nodes remotely
- Ability to detach the machine used to launch a system and have the system continue running
    - Be able to reconnect to the system to monitor and manage nodes running remotely

To achieve our goals, this update will consist of two primary components: an ExecuteRemote action, and a LaunchServiceNode.

The ExecuteRemote action will build off of the refactor he is working on described here: https://github.com/ros2/design/pull/272
Since we don't want to limit remote execution to a single protocol, ExecuteRemote will be a base class and we will create an implementation using SSH which we will call ExecuteRemoteSSH.
Remote machines will be described by a Machine class which will also be a base class for protocol-specific implementations.
Together these components will provide the means to execute processes remotely and should cover most basic use cases.

> The LaunchServiceNode(s) will expose ROS2 services and topics for handling startup and shutdown of processes on remote machines.

Reviewer 2: This LaunchServiceNode could be designed as an independent component - that is possible to use purely locally to detach a launch command from an ongoing launch context - it could then be used by the remote launching functionality.

> This component won't be necessary for basic execution of remote nodes, but would allow remote system to persist beyond the life of the initial LaunchService.

Reviewer 3: I am not clear on this use case, could you elaborate a little bit? especially on the difference from `LifecycleNode` except process statistics?

Reviewer 4: The design is intended to allow you to run a node pretty much like they currently would run, that is, bound to the lifespan of the launch process itself. When the launch process dies, the SSH connection dies, and anything still running on that SSH terminal dies with it.

The `LaunchServiceNode` is intended to work a little bit differently. It would be launched from the SSH terminal, but then detach and run without requiring the SSH connection to remain up. However, by doing that you lose the ability to directly kill a process if you actually want it to stop. Therefore, the `LaunchServiceNode` would provide a way to go back and ask for particular nodes on that server to be started/stopped ad hoc.

This is certainly somewhat similar to some of the functionality envisioned by `LifecycleNode`s, but would allow a way for us to remotely launch and manage nodes which were not natively designed to support the lifecycle feature. Lifecycle would provide nearly all the functionality we would need for this type of behavior, but should the node crash, Lifecycle won't let us bring it back for a full restart, and we also would be lacking other out-of-band control to deal with things such as unresponsive nodes.

(Obviously, these same arguments also apply to the `LaunchServiceNode` itself, so that particular entity should be as clean as possible to reduce the possibility of it getting into an unmanageable state itself.)

Reviewer 2: Question to explicitly state in the doc: Who starts the LaunchServiceNode? How is its lifecycle expected to be managed?

The topics and services will allow remote tools to view and manage the processes running remotely.
To prevent execution of completely arbitrary processes, the node will be given a list of commands it is allowed to run at startup, each with a unique ID assigned by the user.
Requests to start or stop processes would then use the unique ID to indicate to the node which process to start or stop.

## New Classes

#### launch.descriptions.Machine
##### Members
Members would be defined by child classes, since different protocols will require different types of information

#### launch.descriptions.SSHMachine
Contains information needed to run a process on another machine
##### Members
| Name | Type | Description |
|---|---|---|
| hostname   | String | Hostname of the machine. |
| port       | int | Port on to connect to (default 22) |
| ssh\_keys  | String | Path to ssh key file |
| passphrase | String | Optional passphrase if the key uses one. |

#### launch.actions.ExecuteRemote

> Base class for actions that use connection informationn contained in `machine` to start a process on a remote machine.

Reviewer 5: typo: information~n~

#### launch.actions.ExecuteRemoteSSH
Inherits from `launch.actions.ExecuteRemote` and manages the ssh connection
##### Members
Inherited
##### Methods
| Name | Type | Description |
|---|---|---|
| open\_ssh\_connection | SSHConnection | Object for interacting with the SSH connection |
| exec\_remote\_command | void | Executes a command on the remote machine

#### launch.substitutions.RemoteSubstitution
Uses a connection to a machine to run a command remotely and retuns the result as a string
##### Members
##### Constructor
| Argument | Type | Description |
|---|---|---|
| command | Text | The command to run on the remote machine |
| machine | `launch.descriptions.Machine` | An object decribing the machine that the command should be run on. |
##### Methods
| Name | Type | Description |
|---|---|---|
| describe | Text | Override method. Returns a description of this substitution as a string. |
| perform  | Text | Override method. Perform the substitution, given the launch context, and return it as a string. |

## ROS

### Nodes
#### launch\_ros.LaunchServiceNode

> This node runs on each machine in a system is is repsonsible for launching and managing remote processes.

Reviewer 3: nitpick:

```suggestion
This node runs on each machine in a system is is responsible for launching and managing remote processes.
```

Reviewer 5: nit: This node runs on each machine in a system ~is~ it is

#### Params
| Name | Type | Description |
|---|---|---|

> | allowable\_processes | List[Executable] | List of executables the node is allowed to launch |

Reviewer 2: Explicit flag to allow remote launching _any_ process/node? Early in development this could be super useful for testing out various nodes via remote context from developer machine to robot.

Reviewer 2: Or - is it not worth having this allow-list in here at all. Maybe SROS2 takes care of permissions well enough that we don't need to reimplement it. "If you use this you are exposing a vulnerability if you don't use SROS"

Reviewer 2: Perhaps, only allow the LaunchServiceNode to only start known ros nodes on the remote system, rather than arbitrary commands. Restricting to ROS-specific in the remote context, since the local launch can execute arbitrary SSH commands anyways

##### Topics
###### Sent
- heartbeat
###### Received
##### Services
- query\_status
- start\_process
- stop\_process
- shutdown

### Messages

> #### Heartbeat

Reviewer 2: Do all LaunchServiceNodes use a unique topic, or do all publish on the same topic (in which case would need to put machine/service ID in the messages)

Reviewer 2: Maybe this should be configurable, "heartbeat_topic", "process_status_topic"

Contains basic status information and enough information for launch related tools to find the node.
DDS might already provide enough info making this superfluous.
##### Fields
| Name | Type | Description |
|---|---|---|
| timestamp  | `builtin_interface.Timestamp` | Time at which heartbeat message was sent |
| hostname   | String | Hostname of machine sending the status |
| num\_procs | int | Number of processes currently being tracked by the LaunchServiceNode. |

#### ProcessStatus
##### Fields
| Name | Type | Description |
|---|---|---|
| timestamp       | `builtin_interface.Timestamp` | Time at which message was sent |
| process\_name   | String | Name of the process |
| process\_id     | int | A unique numerical identifier for the process. This value is generated by the launch system and should not be expected to be the same as the proc id on the machine. |
| process\_status | int | An enumerated status value. STOPPED=0; RUNNING=1; SUSPENDED=2; |
| mem\_usage      | double | Percent of memory used by the process |
| processor\_load | double | Percent of processor load coming from this process |
| err\_msg        | String | Error string |

> ### Services

Reviewer 2: Noting that each LaunchServiceNode will need to namespace itself to make service calls uniquely-fulfillable

# Multi-Machine and Remote Launching
This article describes the design of the components of the ros2 launch system related to the remote execution of processes.

## Goals
ROS 1's launch system provided the means to launch nodes remotely via SSH by defining a "machine" and referencing that machine in the declaration of each node meant to run on it.
The machine declaration would contain the information necessary to execute processes remotely, such as address, username, password, etc.
The goal of this design is to enable the same ability to launch nodes on remote machines, as well as to provide infrastructure for more advanced capabilities.

- Monitor the status and manage lifecycles of nodes across systems
- Shutdown nodes remotely
- Ability to detach the machine used to launch a system and have the system continue running
    - Be able to reconnect to the system to monitor and manage nodes running remotely

> To achieve our goals, this update will consist of two primary components: an ExecuteRemote action, and a LaunchServiceNode.

Reviewer 2: Based on the re-scope of this design - it seems like this "Goals" section shouldn't be talking about the `LaunchServiceNode` since it's not specified within this design. If I remember correctly, we established that the `LaunchServiceNode` was independently interesting and designable. I do see that we want to talk about how these two things will interact, but I'm not sure if that belongs here in this doc anymore.

> The ExecuteRemote action will provide the basic capaiblity to start processes on remote machines.

Reviewer 2: ```suggestion
The ExecuteRemote action will provide the basic capability to start processes on remote machines.
```

> It will build off of the refactor that is described [here](https://github.com/ros2/design/pull/272) and include a machine object as an additional parameter providing implementation-specific connection information.

Reviewer 2: "Here" is never a very informative link title :)

```suggestion
It will build off of the refactor that is described in [Refactoring ExecuteProcess into Execute and Executable](https://github.com/ros2/design/pull/272) and include a machine object as an additional parameter providing implementation-specific connection information.
```

Since we don't want to limit remote execution to a single protocol, ExecuteRemote will be a base class and we will create an implementation using SSH which we will call ExecuteRemoteSSH.
The machine class will also be a base class for protocol-specific implementations.
Together these components will provide the means to execute processes remotely and should cover most basic use cases.

> The [LaunchServiceNode(s)](154_roslaunch_remote_launch_service_node.md) will expose ROS2 services and topics for handling startup and shutdown of processes on remote machines.

Reviewer 2: It may be easier to evaluate this design tie-in if the linked document was available?

This component won't be necessary for basic execution of remote nodes, but would allow remote system to persist beyond the life of the initial LaunchService.
The topics and services will allow remote tools to view and manage the processes running remotely.
The LaunchServiceNode allows users to monitor and manage the status and lifecycles of nodes across systems, and enables the system to continue running if the machine used to start a system becomes disconnected.

## New Classes

#### launch.descriptions.Machine
##### Members
Members are defined by child classes, since different protocols require different types of information

#### launch.descriptions.SSHMachine
Contains information needed to run a process on another machine via ssh
##### Members
| Name | Type | Description |
|---|---|---|
| hostname   | String | Hostname of the machine. |
| port       | int | Port on the host to connect to (default 22) |

> | ssh\_keys  | String | Path to ssh key file |

Reviewer 2: should this be singular?
```suggestion
| ssh\_key  | String | Path to ssh key file |
```

Author: good catch

| sigkill\_timeout     | List[Substitution] | Time until escalating to `SIGKILL` after `SIGTERM`, as a string or a list of strings and `Substitution`s to be resolved at runtime, defaults to the `LaunchConfiguration` called `sigkill_timeout` |
| emulate\_tty         | boolean | Emulate a tty (terminal), defaults to `False`, but can be overridden with the `LaunchConfiguration` called `emulate_tty`, the value of which is evaluated as true or false according to `evaluate_condition_expression`. |
| prefix               | List[Substitution] | A set of commands/arguments to preceed the `cmd`, used for things like `gdb`/`valgrind` and defaults to the `LaunchConfiguration` called `launch-prefix` |
| output               | ?? | Configuration for process output logging. Defaults to `log` i.e. log both `stdout` and `stderr` to launch main log file and stderr to the screen. |
| output\_format       | ?? | For logging each output line, supporting `str.format()` substitutions with the following keys in scope: `line` to reference the raw output line and `this` to reference this action instance. |

> | log\_cmd             | boolean | If `True`, prints the final cmd before executing the process, which is useful for debugging when substitutions are involved. |

Reviewer 2: maybe should have a note that this will avoid printing any key passphrases or other sensitive information

Author: Oh, good point, I will include a note about that.

| on\_exit             | List[LaunchDescriptionEntity] | List of actions to execute upon process exit.|
| persistent\_connection | boolean | Whether the LaunchService should maintain a persistent connection to the ssh instance |

> | machine              | `launch.descriptions.SSHMachine` | The machine on which to execute the process |

Reviewer 2: Want this to take the abstract interface, right?

```suggestion
| machine              | `launch.descriptions.Machine` | The machine on which to execute the process |
```

Author: You are correct, thanks
